core/deployment/src/main/java/io/quarkus/deployment/util/GenericTypeUtil.java

```diff
@@ -0,0 +1,160 @@
+
+                String unmatchedParameter = argument.asTypeVariable().identifier();
+
+                for (Level level : findResult.levels) {
+                    Type matchingCapturedType = null;
+                    for (int j = 0; j < level.definitions.size(); j++) {
+                        final Type definition = level.definitions.get(j);
+                        if ((definition instanceof TypeVariable)
+                                && unmatchedParameter.equals(definition.asTypeVariable().identifier())) {
+                            matchingCapturedType = level.captures.get(j);
+                            break; // out of the definitions loop
+                        }
+                    }
+                    // at this point their MUST be a match, if there isn't we have made some mistake in the implementation
+                    if (matchingCapturedType == null) {
+                        throw new IllegalStateException("Error retrieving generic types");
+                    }
+                    if (isDirectlyHandledType(matchingCapturedType)) {
+                        result.add(matchingCapturedType);
+                        break; // out of level loop
+                    }
+                    if (matchingCapturedType instanceof TypeVariable) {
+                        // continue the search in the lower levels using the new name
+                        unmatchedParameter = matchingCapturedType.asTypeVariable().identifier();
+                    }
+                }
+            }
+        }
+
+        if (result.size() != findResult.argumentsOfMatch.size()) {
+            throw new IllegalStateException("Unable to properly match generic types");
+        }
+
+        return result;
+    }
+
+    private static boolean isDirectlyHandledType(Type matchingCapturedType) {
+        return (matchingCapturedType instanceof ClassType) ||
+                (matchingCapturedType instanceof ParameterizedType);
+    }
+
+    private static FindResult find(DotName iface, ClassInfo classInfo, IndexView index, List<Level> visitedTypes) {
+        final List<Type> interfaceTypes = classInfo.interfaceTypes();
+        final List<Type> otherInterfaces = new ArrayList<>();
+        for (Type interfaceType : interfaceTypes) {
+            if (iface.equals(interfaceType.name())) {
+                if (interfaceType instanceof ParameterizedType) {
+                    return new FindResult(interfaceType.asParameterizedType().arguments(), visitedTypes);
+                }
+                return null;
+            }
+            // we only check the other interfaces if we didn't find anything
+            // this ensures that we also use the "closest" path to the target interface
+            otherInterfaces.add(interfaceType);
+        }
+
+        for (Type otherInterface : otherInterfaces) {
+            final FindResult findResult = find(iface, index.getClassByName(otherInterface.name()), index,
```

Comment:
Are you sure this one is in the index? There is a similar issue a few lines below.

Reply:
Yeah... As mentioned above the entire hierarchy needs to be in the index, there is no way around that as far as I can tell.

So perhaps I should add a check here and throw an exception if it's not

Reply:
Added index check here as well